File: openstudio-hvac/lib/openstudio_hvac/systems/baseboards.rb
module OpenStudioHVAC
  module Systems
    # Zone baseboards (electric or hot-water), ported from openstudio-standards
    # add_zone_baseboards.
    module Baseboards
      # @param model [OpenStudio::Model::Model]
      # @param zone [OpenStudio::Model::ThermalZone]
      # @param baseboard_type [String] 'Electric' or 'Hot Water'
      # @param hw_loop [OpenStudio::Model::PlantLoop, nil] required for 'Hot Water'
      # @return [void]
      def self.add(model, zone, baseboard_type:, hw_loop: nil)
        case baseboard_type
        when 'Electric'
          baseboard = OpenStudio::Model::ZoneHVACBaseboardConvectiveElectric.new(model)
          baseboard.addToThermalZone(zone)
        when 'Hot Water', 'HotWater'
          raise(ArgumentError, 'a hot water loop is required for Hot Water baseboards') if hw_loop.nil?

          coil = OpenStudio::Model::CoilHeatingWaterBaseboard.new(model)
          hw_loop.addDemandBranchForComponent(coil)
          baseboard = OpenStudio::Model::ZoneHVACBaseboardConvectiveWater.new(
            model, model.alwaysOnDiscreteSchedule, coil
          )
          baseboard.addToThermalZone(zone)
        when 'None', nil
          # no baseboards
        else
          raise(ArgumentError, "'#{baseboard_type}' is not a valid baseboard type")
        end
      end
    end
  end
end

File: python/btap/modeling/hvac/systems/baseboards.py
"""Zone baseboards (electric or hot-water), ported from openstudio-standards
add_zone_baseboards."""

from __future__ import annotations

import openstudio


def add(model, zone, baseboard_type, hw_loop=None):
    """Add a baseboard to a thermal zone.

    :param model: openstudio.model.Model
    :param zone: openstudio.model.ThermalZone
    :param baseboard_type: 'Electric' or 'Hot Water'
    :param hw_loop: openstudio.model.PlantLoop or None (required for 'Hot Water')
    :return: None
    """
    if baseboard_type == 'Electric':
        baseboard = openstudio.model.ZoneHVACBaseboardConvectiveElectric(model)
        baseboard.addToThermalZone(zone)
    elif baseboard_type in ('Hot Water', 'HotWater'):
        if hw_loop is None:
            raise ValueError('a hot water loop is required for Hot Water baseboards')

        coil = openstudio.model.CoilHeatingWaterBaseboard(model)
        hw_loop.addDemandBranchForComponent(coil)
        baseboard = openstudio.model.ZoneHVACBaseboardConvectiveWater(
            model, model.alwaysOnDiscreteSchedule(), coil)
        baseboard.addToThermalZone(zone)
    elif baseboard_type in ('None', None):
        pass  # no baseboards
    else:
        raise ValueError(f"'{baseboard_type}' is not a valid baseboard type")
